src/expr/src/table_function/mod.rs

```diff
@@ -62,3 +73,184 @@ pub fn build_from_prost(prost: &TableFunctionProst) -> Result<BoxedTableFunction
+        }
+    }
+
+    pub fn return_type(&self) -> DataType {
+        match self {
+            ProjectSetSelectItem::TableFunction(tf) => tf.return_type(),
+            ProjectSetSelectItem::Expr(expr) => expr.return_type(),
+        }
+    }
+
+    fn eval(&self, input: &DataChunk) -> Result<ProjectSetSelectItemResult> {
+        match self {
+            ProjectSetSelectItem::TableFunction(tf) => tf
+                .eval(input)
+                .map(ProjectSetSelectItemResult::TableFunction),
+            ProjectSetSelectItem::Expr(expr) => {
+                expr.eval(input).map(ProjectSetSelectItemResult::Expr)
+            }
+        }
+    }
+
+    /// First column will be `projected_row_id`, which represents the index in the output table
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    pub async fn execute(
+        select_list: Arc<Vec<Self>>,
+        data_types: Vec<DataType>,
+        input: &DataChunk,
+    ) {
+        assert!(!select_list.is_empty());
+
+        let results: Vec<_> = select_list
+            .iter()
+            .map(|select_item| select_item.eval(input))
+            .try_collect()?;
+
+        let mut iters = results.iter().map(|r| r.iter()).collect_vec();
+
+        // each iteration corresponds to the outputs of one input row
+        loop {
+            let items = iters.iter_mut().map(|iter| iter.next()).collect_vec();
+
+            if items[0].is_none() {
+                assert!(items.iter().all(|i| i.is_none()));
+                break;
+            }
+            let items = items.into_iter().map(|i| i.unwrap()).collect_vec();
+
+            // The maximum length of the results of table functions will be the output length.
+            let max_tf_len = items
+                .iter()
+                .map(|i| i.as_ref().map_left(|arr| arr.len()).left_or(0))
+                .max()
+                .unwrap();
+            let builders = data_types
+                .clone()
+                .into_iter()
+                .map(|ty| ty.create_array_builder(max_tf_len));
+
+            let mut columns = vec![index_array_column(max_tf_len)];
+
+            for (item, mut builder) in items.into_iter().zip_eq(builders) {
+                match item {
+                    Either::Left(array_ref) => {
+                        builder.append_array(&array_ref)?;
+                        for _ in 0..(max_tf_len - array_ref.len()) {
+                            builder.append_null()?;
+                        }
+                    }
+                    Either::Right(datum_ref) => {
+                        for _ in 0..max_tf_len {
+                            builder.append_datum_ref(datum_ref)?;
+                        }
+                    }
+                }
+                let array = builder.finish()?;
+                columns.push(Column::new(Arc::new(array)));
+            }
+            let chunk = DataChunk::new(columns, max_tf_len);
+
+            yield chunk;
```

Comment:
returns one big chunk is better? 🤔

Reply:
Just found another problem: For batch project set, consolidate to a large chunk is easy enough. Buf for stream project set, we need to replicate `op` for each row...

So it seems that we can't reuse code between batch & stream 🥵